Prevent concurrent modification of keysWithEvents map

dwijnand opened the pull request:
Fixup and rebase of @mechkg's #89.

Fixes #86
which means
Fixes sbt/sbt#3687
Fixes sbt/sbt#3695
Fixes sbt/sbt#3775

mechkg committed:
Prevent concurrent modification of keysWithEvents map

Fixes #86
which means
Fixes sbt/sbt#3687
Fixes sbt/sbt#3695
Fixes sbt/sbt#3775

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?io=pull/90/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

eed3si9n commented:
```
[info] PollingWatchServiceSpec:
[info] - should detect modified files *** FAILED ***
[info]   false was not equal to true (SourceModificationWatchSpec.scala:322)
```

Should we be worried?

dwijnand committed:
Really make "should detect modified files" pending

I don't understand how you're meant to do pendingUntilFixed in ScalaTest, like specs2..

ches commented on io/src/test/scala/sbt/internal/io/SourceModificationWatchSpec.scala:
```diff
@@ -21,9 +21,9 @@ abstract class SourceModificationWatchSpec(
+    // watchTest(parentDir)(pollDelay, maxWait) {
+    //   IO.write(file, "bar")
+    // }
     pending // until fixed https://github.com/sbt/io/issues/82
```
You wrap it as a block I think 😊 

```scala
pendingUntilFixed {
  watchTest(parentDir)(pollDelay, maxWait) {
    IO.write(file, "bar")
  }
}

dwijnand replied:
Ah! Cheers.
